docs/01-app/02-guides/package-bundling.mdx

```diff
@@ -70,69 +136,144 @@ const nextConfig = {
+
+### Heavy client workloads
+
+A common cause of large client bundles is doing expensive rendering work in Client Components. This often happens with libraries that exist only to transform data into UI, such as syntax highlighting, chart rendering, or markdown parsing.
+
+If that work does not require browser APIs or user interaction, it can be run in a Server Component.
+
+In this example, a prism based highlighter runs in a Client Component. Even though the final output is just a `<code>` block, the entire highlighting library is bundled into the client JavaScript bundle:
+
+```tsx filename="app/blog/[slug]/page.tsx"
+'use client'
+
+import Highlight from 'prism-react-renderer'
+import theme from 'prism-react-renderer/themes/github'
+
+export default function Page() {
+  const code = `export function hello() {
+    console.log("hi")
+  }`
+
+  return (
+    <article>
+      <h1>Blog Post Title</h1>
+
+      {/* The prism package and its tokenization logic are shipped to the client */}
+      <Highlight code={code} language="tsx" theme={theme}>
+        {({ className, style, tokens, getLineProps, getTokenProps }) => (
+          <pre className={className} style={style}>
+            <code>
+              {tokens.map((line, i) => (
+                <div key={i} {...getLineProps({ line })}>
+                  {line.map((token, key) => (
+                    <span key={key} {...getTokenProps({ token })} />
+                  ))}
+                </div>
+              ))}
+            </code>
+          </pre>
+        )}
+      </Highlight>
+    </article>
+  )
+}
+```
 
-<PagesOnly>
+This increases bundle size because the client must download and execute the highlighting library, even though the result is static HTML.
 
-## Bundling specific packages
+Instead, move the highlighting logic to a Server Component and render the final HTML on the server. The client will only receive the rendered markup.
 
-To bundle specific packages, you can use the [`transpilePackages`](/docs/app/api-reference/config/next-config-js/transpilePackages) option in your `next.config.js`. This option is useful for bundling external packages that are not pre-bundled, for example, in a monorepo or imported from `node_modules`.
+```tsx filename="app/blog/[slug]/page.tsx"
+import { codeToHtml } from 'shiki'
+
+export default async function Page() {
+  const code = `export function hello() {
+    console.log("hi")
+  }`
+
+  // The Shiki package runs on the server and is never bundled for the client.
+  const highlightedHtml = await codeToHtml(code, {
+    lang: 'tsx',
+    theme: 'github-dark',
+  })
+
+  return (
+    <article>
+      <h1>{post.title}</h1>
+
+      {/* Client receives plain markup */}
+      <pre>
+        <code dangerouslySetInnerHTML={{ __html: highlightedHtml }} />
```

Comment:
Nit: not related to the bundle analyzer, but is `dangerouslySetInnerHTML` with shiki always safe?

Reply:
Yeah, I know, not the best, but used in the [Shiki docs](https://shiki.style/packages/next#:~:text=return%20%3Cdiv%20dangerouslySetInnerHTML%3D%7B%7B%20__html%3A%20out%20%7D%7D%20/%3E).